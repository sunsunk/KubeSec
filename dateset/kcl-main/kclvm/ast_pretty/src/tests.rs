use std::path::{Path, PathBuf};

use super::print_ast_module;
use kclvm_parser::parse_file_force_errors;
use pretty_assertions::assert_eq;

const FILE_INPUT_SUFFIX: &str = ".input";
const FILE_OUTPUT_SUFFIX: &str = ".output";
const TEST_CASES: &[&str] = &[
    "arguments",
    "empty",
    "if_stmt",
    "import",
    "unary",
    "codelayout",
    "collection_if",
    "comment",
    "index_sign",
    "joined_str",
    "lambda",
    "orelse",
    "quant",
    "rule",
    "str",
    "type_alias",
    "unification",
];

fn read_data(data_name: &str) -> (String, String) {
    let mut filename = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_INPUT_SUFFIX))
            .display()
            .to_string(),
    );

    let module = parse_file_force_errors(filename.to_str().unwrap(), None);

    let mut filename_expect = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    filename_expect.push(
        Path::new("src")
            .join("test_data")
            .join(format!("{}{}", data_name, FILE_OUTPUT_SUFFIX))
            .display()
            .to_string(),
    );
    (
        print_ast_module(&module.unwrap()),
        std::fs::read_to_string(filename_expect.to_str().unwrap()).unwrap(),
    )
}

#[test]
fn test_ast_printer() {
    for case in TEST_CASES {
        let (data_input, data_output) = read_data(case);

        #[cfg(target_os = "windows")]
        let data_output = data_output.replace("\r\n", "\n");

        assert_eq!(data_input, data_output, "Test failed on {}", case);
    }
}
