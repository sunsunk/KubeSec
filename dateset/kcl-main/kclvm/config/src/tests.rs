use kclvm_version as version;
use std::{
    collections::HashMap,
    env,
    fs::{self, File},
    path::{Path, PathBuf},
};

use crate::{
    cache::{load_pkg_cache, save_pkg_cache, CacheOption},
    modfile::{get_vendor_home, KCL_PKG_PATH},
};

#[test]
fn test_vendor_home() {
    env::set_var(KCL_PKG_PATH, "test_vendor_home");
    assert_eq!(get_vendor_home(), "test_vendor_home");
    env::remove_var(KCL_PKG_PATH);

    #[cfg(target_os = "windows")]
    let root_dir = env::var("USERPROFILE").unwrap();
    #[cfg(not(target_os = "windows"))]
    let root_dir = env::var("HOME").unwrap();

    let kpm_home = PathBuf::from(root_dir)
        .join(".kcl")
        .join("kpm")
        .canonicalize()
        .unwrap();
    assert_eq!(get_vendor_home(), kpm_home.display().to_string())
}

#[test]
fn test_pkg_cache() {
    let root = PathBuf::from("./src/testdata/test_cache/")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    let mut external_pkgs = HashMap::new();
    external_pkgs.insert(
        "test_vendor".to_string(),
        "./src/testdata/test_vendor".to_string(),
    );

    let lock_path = Path::new(&root)
        .join(".kclvm/cache")
        .join(format!("{}-{}", version::VERSION, version::CHECK_SUM))
        .join("test_target");

    fs::create_dir_all(lock_path.clone()).unwrap();
    File::create(lock_path.join("test_vendor.lock")).unwrap();

    save_pkg_cache(
        &root,
        "test_target",
        "test_vendor",
        "test_data",
        CacheOption::default(),
        &external_pkgs,
    )
    .unwrap();

    assert_eq!(
        load_pkg_cache(
            &root,
            "test_target",
            "test_vendor",
            CacheOption::default(),
            &external_pkgs,
        ),
        Some("test_data".to_string())
    )
}
