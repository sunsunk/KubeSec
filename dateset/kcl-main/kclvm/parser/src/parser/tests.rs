use crate::parse_file_force_errors;
use regex::Regex;

#[test]
fn test_parse_file_not_found() {
    match parse_file_force_errors("The file path is invalid", None) {
        Ok(_) => {
            panic!("unreachable")
        }
        Err(err_msg) => {
            assert!(
                Regex::new(r"^Failed to load KCL file 'The file path is invalid'. Because.*")
                    .unwrap()
                    .is_match(&err_msg.to_string())
            );
        }
    }
}
