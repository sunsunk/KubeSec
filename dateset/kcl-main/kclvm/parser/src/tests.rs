use std::{
    env,
    panic::{catch_unwind, set_hook},
};

use compiler_base_span::{FilePathMapping, SourceMap};
use kclvm_config::modfile::{get_vendor_home, KCL_PKG_PATH};

use crate::*;

use core::any::Any;

mod ast;
mod error_recovery;
mod expr;
mod file;
mod types;

#[macro_export]
macro_rules! parse_expr_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_expr_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_module_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_module_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_type_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_type_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_type_node_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_type_node_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_file_ast_json_snapshot {
    ($name:ident, $filename:expr, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_file_ast_json($filename, $src));
        }
    };
}

#[macro_export]
macro_rules! parse_file_snapshot {
    ($name:ident, $filename:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_file_string($filename));
        }
    };
}

pub(crate) fn parsing_expr_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let expr = parser.parse_expr();
            format!("{expr:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub(crate) fn parsing_module_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let module = parser.parse_module();
            format!("{module:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub(crate) fn parsing_type_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    create_session_globals_then(|| {
        let stream = parse_token_streams(sess, src, new_byte_pos(0));
        let mut parser = Parser::new(sess, stream);
        let typ = parser.parse_type_annotation();
        format!("{typ:#?}\n")
    })
}

pub(crate) fn parsing_type_node_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    create_session_globals_then(|| {
        let stream = parse_token_streams(sess, src, new_byte_pos(0));
        let mut parser = Parser::new(sess, stream);
        let typ = parser.parse_type_annotation();
        typ.node.to_string()
    })
}

pub(crate) fn parsing_file_ast_json(filename: &str, src: &str) -> String {
    let m = crate::parse_file_with_global_session(
        Arc::new(ParseSession::default()),
        filename,
        Some(src.into()),
    )
    .unwrap();
    serde_json::ser::to_string_pretty(&m).unwrap()
}

pub(crate) fn parsing_file_string(filename: &str) -> String {
    let code = std::fs::read_to_string(filename).unwrap();
    let m = crate::parse_file(filename.trim_start_matches("testdata/"), Some(code))
        .expect(filename)
        .module;
    serde_json::ser::to_string_pretty(&m).unwrap()
}

pub fn check_result_panic_info(result: Result<(), Box<dyn Any + Send>>) {
    if let Err(e) = result {
        assert!(e.downcast::<String>().is_ok());
    };
}

const PARSE_EXPR_INVALID_TEST_CASES: &[&str] =
    &["fs1_i1re1~s", "fh==-h==-", "8_________i", "1MM", "0x00x"];

#[test]
pub fn test_parse_expr_invalid() {
    for case in PARSE_EXPR_INVALID_TEST_CASES {
        set_hook(Box::new(|_| {}));
        let result = catch_unwind(|| {
            parse_expr(case);
        });
        check_result_panic_info(result);
    }
}

const PARSE_FILE_INVALID_TEST_CASES: &[&str] = &[
    "a: int",                   // No initial value error
    "a -",                      // Invalid binary expression error
    "a?: int",                  // Invalid optional annotation error
    "if a not is not b: a = 1", // Logic operator error
    "if True:\n  a=1\n b=2",    // Indent error with recovery
    "a[1::::]",                 // List slice error
    "a[1 a]",                   // List index error
    "{a ++ 1}",                 // Config attribute operator error
    "func(a=1,b)",              // Call argument error
    "'${}'",                    // Empty string interpolation error
    "'${a: jso}'",              // Invalid string interpolation format spec error
];

#[test]
pub fn test_parse_file_invalid() {
    for case in PARSE_FILE_INVALID_TEST_CASES {
        let result = parse_file_force_errors("test.k", Some((&case).to_string()));
        assert!(result.is_err(), "case: {case}, result {result:?}");
    }
}

pub fn test_vendor_home() {
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCL_PKG_PATH, vendor);
    assert_eq!(get_vendor_home(), vendor.to_string());
}

fn set_vendor_home() -> String {
    // set env vendor
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCL_PKG_PATH, vendor);
    debug_assert_eq!(get_vendor_home(), vendor.to_string());
    vendor.to_string()
}

#[test]
/// The testing will set environment variables,
/// so can not to execute test cases concurrently.
fn test_in_order() {
    test_import_vendor_by_external_arguments();
    println!("{:?} PASS", "test_import_vendor_by_external_arguments");
    test_import_vendor_without_vendor_home();
    println!("{:?} PASS", "test_import_vendor_without_vendor_home");
    test_import_vendor_without_kclmod();
    println!("{:?} PASS", "test_import_vendor_without_kclmod");
    test_import_vendor();
    println!("{:?} PASS", "test_import_vendor");
    test_import_vendor_with_same_internal_pkg();
    println!("{:?} PASS", "test_import_vendor_with_same_internal_pkg");
    test_import_vendor_without_kclmod_and_same_name();
    println!(
        "{:?} PASS",
        "test_import_vendor_without_kclmod_and_same_name"
    );
    test_vendor_home();
    println!("{:?} PASS", "test_vendor_home");
    test_pkg_not_found_suggestion();
    println!("{:?} PASS", "test_pkg_not_found_suggestion");
}

pub fn test_import_vendor() {
    let module_cache = KCLModuleCache::default();
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![
        ("assign.k", vec!["__main__", "assign", "assign.assign"]),
        (
            "config_expr.k",
            vec!["__main__", "config_expr", "config_expr.config_expr_02"],
        ),
        (
            "nested_vendor.k",
            vec![
                "__main__",
                "nested_vendor",
                "nested_vendor.nested_vendor",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
        (
            "subpkg.k",
            vec![
                "__main__",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub",
            ],
        ),
    ];

    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();

    let test_fn =
        |test_case_name: &&str, pkgs: &Vec<&str>, module_cache: Option<KCLModuleCache>| {
            let test_case_path = dir.join(test_case_name).display().to_string();
            let m = load_program(sess.clone(), &[&test_case_path], None, module_cache)
                .unwrap()
                .program;
            assert_eq!(m.pkgs.len(), pkgs.len());
            m.pkgs.into_iter().for_each(|(name, modules)| {
                println!("{:?} - {:?}", test_case_name, name);
                assert!(pkgs.contains(&name.as_str()));
                for pkg in pkgs.clone() {
                    if name == pkg {
                        if name == "__main__" {
                            assert_eq!(modules.len(), 1);
                            assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                        } else {
                            modules.into_iter().for_each(|module| {
                                assert!(module.filename.contains(&vendor));
                            });
                        }
                        break;
                    }
                }
            });
        };

    test_cases
        .iter()
        .for_each(|(test_case_name, pkgs)| test_fn(test_case_name, pkgs, None));

    test_cases.iter().for_each(|(test_case_name, pkgs)| {
        test_fn(test_case_name, pkgs, Some(module_cache.clone()))
    });
}

pub fn test_import_vendor_without_kclmod() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![("import_vendor.k", vec!["__main__", "assign.assign"])];

    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .canonicalize()
        .unwrap();

    test_cases.into_iter().for_each(|(test_case_name, pkgs)| {
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None, None)
            .unwrap()
            .program;
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    });
}

pub fn test_import_vendor_without_vendor_home() {
    env::set_var(KCL_PKG_PATH, "");
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath assign not found in the program",
                "try 'kcl mod add assign' to download the package not found",
                "find more package on 'https://artifacthub.io'",
                "pkgpath assign.assign not found in the program",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }

    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath assign not found in the program",
                "try 'kcl mod add assign' to download the package not found",
                "find more package on 'https://artifacthub.io'",
                "pkgpath assign.assign not found in the program",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_with_same_internal_pkg() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("same_name.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `same_vendor` is found multiple times in the current package and vendor package"
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `same_vendor` is found multiple times in the current package and vendor package"
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_without_kclmod_and_same_name() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .join("same_name")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `assign` is found multiple times in the current package and vendor package",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }

    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `assign` is found multiple times in the current package and vendor package",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_by_external_arguments() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let module_cache = KCLModuleCache::default();
    let external_dir = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap();

    let test_cases = vec![
        (
            "import_by_external_assign.k",
            "assign",
            vec!["__main__", "assign"],
        ),
        (
            "import_by_external_config_expr.k",
            "config_expr",
            vec!["__main__", "config_expr"],
        ),
        (
            "import_by_external_nested_vendor.k",
            "nested_vendor",
            vec![
                "__main__",
                "nested_vendor",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
        (
            "import_by_external_vendor_subpkg.k",
            "vendor_subpkg",
            vec![
                "__main__",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
    ];

    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .canonicalize()
        .unwrap();

    let test_fn = |test_case_name: &&str,
                   dep_name: &&str,
                   pkgs: &Vec<&str>,
                   module_cache: Option<KCLModuleCache>| {
        let mut opts = LoadProgramOptions::default();
        opts.package_maps.insert(
            dep_name.to_string(),
            external_dir.join(dep_name).display().to_string(),
        );
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None, module_cache)
            .unwrap()
            .program;
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    };

    test_cases
        .iter()
        .for_each(|(test_case_name, dep_name, pkgs)| test_fn(test_case_name, dep_name, pkgs, None));

    test_cases
        .iter()
        .for_each(|(test_case_name, dep_name, pkgs)| {
            test_fn(test_case_name, dep_name, pkgs, Some(module_cache.clone()))
        });
}

#[test]
fn test_get_compile_entries_from_paths() {
    let testpath = PathBuf::from("./src/testdata/multimods")
        .canonicalize()
        .unwrap();

    // [`kcl1_path`] is a normal path of the package [`kcl1`] root directory.
    // It looks like `/xxx/xxx/xxx`.
    let kcl1_path = testpath.join("kcl1");

    // [`kcl2_path`] is a mod relative path of the packege [`kcl2`] root directory.
    // It looks like `${kcl2:KCL_MOD}/xxx/xxx`
    let kcl2_path = PathBuf::from("${kcl2:KCL_MOD}/main.k");

    // [`kcl3_path`] is a mod relative path of the [`__main__`] packege.
    let kcl3_path = PathBuf::from("${KCL_MOD}/main.k");

    // [`package_maps`] is a map to show the real path of the mod relative path [`kcl2`].
    let mut opts = LoadProgramOptions::default();
    opts.package_maps.insert(
        "kcl2".to_string(),
        testpath.join("kcl2").to_str().unwrap().to_string(),
    );

    // [`get_compile_entries_from_paths`] will return the map of package name to package root real path.
    let entries = get_compile_entries_from_paths(
        &[
            kcl1_path.to_str().unwrap().to_string(),
            kcl2_path.display().to_string(),
            kcl3_path.display().to_string(),
        ],
        &opts,
    )
    .unwrap();

    assert_eq!(entries.len(), 3);

    assert_eq!(entries.get_nth_entry(0).unwrap().name(), "__main__");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(0).unwrap().path())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        kcl1_path.canonicalize().unwrap().to_str().unwrap()
    );

    assert_eq!(entries.get_nth_entry(1).unwrap().name(), "kcl2");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(1).unwrap().path())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        testpath
            .join("kcl2")
            .canonicalize()
            .unwrap()
            .to_str()
            .unwrap()
    );

    assert_eq!(entries.get_nth_entry(2).unwrap().name(), "__main__");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(2).unwrap().path())
            .canonicalize()
            .unwrap()
            .to_str()
            .unwrap(),
        kcl1_path.canonicalize().unwrap().to_str().unwrap()
    );
}

#[test]
fn test_dir_with_k_code_list() {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let testpath = PathBuf::from("./src/testdata/test_k_code_list")
        .canonicalize()
        .unwrap();

    let mut opts = LoadProgramOptions::default();
    opts.k_code_list = vec!["test_code = 1".to_string()];

    match load_program(
        sess.clone(),
        &[&testpath.display().to_string()],
        Some(opts.clone()),
        None,
    ) {
        Ok(_) => panic!("unreachable code"),
        Err(err) => assert!(err.to_string().contains("Invalid code list")),
    }

    match load_program(
        sess.clone(),
        &[&testpath.display().to_string()],
        Some(opts),
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => panic!("unreachable code"),
        Err(err) => assert!(err.to_string().contains("Invalid code list")),
    }
}

pub fn test_pkg_not_found_suggestion() {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from("./src/testdata/pkg_not_found")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("suggestions.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath k9s not found in the program",
                "try 'kcl mod add k9s' to download the package not found",
                "find more package on 'https://artifacthub.io'",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}
