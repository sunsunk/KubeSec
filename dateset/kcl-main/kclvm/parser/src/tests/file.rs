use crate::tests::parse_file_snapshot;

parse_file_snapshot!(assert_1, "testdata/assert-01.k");
parse_file_snapshot!(assert_2, "testdata/assert-02.k");
parse_file_snapshot!(assert_3, "testdata/assert-03.k");
parse_file_snapshot!(assert_if_0, "testdata/assert-if-0.k");
parse_file_snapshot!(assert_if_1, "testdata/assert-if-1.k");
parse_file_snapshot!(assert_if_2, "testdata/assert-if-2.k");
parse_file_snapshot!(assign_1, "testdata/assign-01.k");
parse_file_snapshot!(config_expr_1, "testdata/config_expr-01.k");
parse_file_snapshot!(config_expr_2, "testdata/config_expr-02.k");
parse_file_snapshot!(config_expr_3, "testdata/config_expr-03.k");
parse_file_snapshot!(config_expr_4, "testdata/config_expr-04.k");
parse_file_snapshot!(import_1, "testdata/import-01.k");
parse_file_snapshot!(if_1, "testdata/if-01.k");
parse_file_snapshot!(if_2, "testdata/if-02.k");
parse_file_snapshot!(if_3, "testdata/if-03.k");
parse_file_snapshot!(type_1, "testdata/type-01.k");
parse_file_snapshot!(hello_win, "testdata/hello_win.k");
