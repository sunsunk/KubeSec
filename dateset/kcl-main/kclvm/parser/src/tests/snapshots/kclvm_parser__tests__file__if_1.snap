---
source: parser/src/tests/file.rs
expression: "crate::tests::parsing_file_string(\"testdata/if-01.k\")"
---
{
  "filename": "if-01.k",
  "pkg": "__main__",
  "doc": null,
  "name": "__main__",
  "body": [
    {
      "node": {
        "type": "Assign",
        "targets": [
          {
            "node": {
              "names": [
                {
                  "node": "a",
                  "filename": "if-01.k",
                  "line": 1,
                  "column": 0,
                  "end_line": 1,
                  "end_column": 1
                }
              ],
              "pkgpath": "",
              "ctx": "Store"
            },
            "filename": "if-01.k",
            "line": 1,
            "column": 0,
            "end_line": 1,
            "end_column": 1
          }
        ],
        "value": {
          "node": {
            "type": "NumberLit",
            "binary_suffix": null,
            "value": {
              "type": "Int",
              "value": 1
            }
          },
          "filename": "if-01.k",
          "line": 1,
          "column": 4,
          "end_line": 1,
          "end_column": 5
        },
        "ty": null
      },
      "filename": "if-01.k",
      "line": 1,
      "column": 0,
      "end_line": 1,
      "end_column": 5
    },
    {
      "node": {
        "type": "If",
        "body": [
          {
            "node": {
              "type": "Assign",
              "targets": [
                {
                  "node": {
                    "names": [
                      {
                        "node": "bbb",
                        "filename": "if-01.k",
                        "line": 4,
                        "column": 4,
                        "end_line": 4,
                        "end_column": 7
                      }
                    ],
                    "pkgpath": "",
                    "ctx": "Store"
                  },
                  "filename": "if-01.k",
                  "line": 4,
                  "column": 4,
                  "end_line": 4,
                  "end_column": 7
                }
              ],
              "value": {
                "node": {
                  "type": "NumberLit",
                  "binary_suffix": null,
                  "value": {
                    "type": "Int",
                    "value": 2
                  }
                },
                "filename": "if-01.k",
                "line": 4,
                "column": 10,
                "end_line": 4,
                "end_column": 11
              },
              "ty": null
            },
            "filename": "if-01.k",
            "line": 4,
            "column": 4,
            "end_line": 4,
            "end_column": 11
          }
        ],
        "cond": {
          "node": {
            "type": "Identifier",
            "names": [
              {
                "node": "a",
                "filename": "if-01.k",
                "line": 3,
                "column": 3,
                "end_line": 3,
                "end_column": 4
              }
            ],
            "pkgpath": "",
            "ctx": "Load"
          },
          "filename": "if-01.k",
          "line": 3,
          "column": 3,
          "end_line": 3,
          "end_column": 4
        },
        "orelse": []
      },
      "filename": "if-01.k",
      "line": 3,
      "column": 0,
      "end_line": 4,
      "end_column": 12
    }
  ],
  "comments": []
}
